doc/explanation/package-management.md

```diff
@@ -0,0 +1,220 @@
+there is no obligation to switch to declaring dependencies in the
+`dune-project`. Likewise the generation of `.opam` files will still work.
+
+## Locking
+
+To go from a set of dependency constraints for a project to a set of installed
+packages and versions there needs to be a step to determine the right packages
+and their versions to be installed.
+
+In `opam` this process happens as part of `opam install`, thus linking finding
+a solution that satifies the given constrains and installation into one step.
+Dune on the other hand moves finding a solution and installing into separate
+steps. First a solution is found and then packages are installed separately.
+The idea of finding a solution and recording it for later is popular in many
+package managers and is usually called locking.
+
+:::{note}
+`opam` also supports creating lock-files. However these are not as central to
+the opam workflow as they are in the case of package management in Dune, which
+always requires a set of locked packages.
+:::
+
+A package lock is in the most general sense just a set of specific packages and
+their versions to be installed.
+
+A Dune lockfile extends this to a directory with files that describe the
+dependencies. It includes the name of the package and the version. Unlike many
+other package managers, the files include a lot of other information as well,
+such as the location of the source archives to download (since there is no
+central location for all archives), the build instructions (since each package
+can use its own way of building) as well as additional metadata like the system
+packages it depends upon.
+
+The files are stored in a directory as separate files, because that makes it
+easier to manage them in source control (less potential conflicts). Storing
+additional files like patches is also simpler this way.
+
+### Package Repository Management
+
+To find a valid solution that allows a project to be built it is necessary to
+know what packages exist, what versions of these packages exist, what other
+packages these depend on etc.
+
+In OPAM this information is tracked in a central repository called
+[OPAM-repository](https://github.com/ocaml/opam-repository) which contains all
+the metadata for published packages.
+
+It is managed using Git and OPAM typically uses a snapshot of it to look up the
+dependencies when trying to find a solution that satisfies the constraints.
+
+Likewise, Dune uses the same repository, however instead of snapshots of the
+files it uses the Git repository directly. In fact, Dune maintains a shared
+internal cache containing all Git repositories that projects use. This has the
+advantage that updates to the metadata are very fast since only the newest
+revisions have to be retrieved. The downside is that for the creation of the
+cache the entire repository has to be cloned first.
+
+Given a priorization of fast updates, whenever Dune needs to determine the
```

Comment:
Does it mean `dune pkg lock` always needs network access?

Reply:
It needs network access unless the repos you use are pinned to a revision that's in the rev store.

Reply:
Do you have a way to control this? Ie. what if I want to rebuild a package in a train (with all the dependencies already cached locally because I've built the project in the past)

Reply:
If you have built the project in the past, then you don't need to relock and it works offline (as fetch rules are cached out of the box since #10710), even if you remove `_build`.

If you want to relock without internet it requires a bit if configuration. You can define the opam-repos to use for locking to use an exact revision [like here](https://dune.readthedocs.io/en/latest/tutorials/dune-package-management/repos.html), in which case Dune will first check whether the revision is available in the revision cache, and only pull if the revision was not found yet.